File: app/Services/SupporterMessageViewService.php
<?php
// Copyright © 2024–2026 Jasper Ford
// SPDX-License-Identifier: AGPL-3.0-or-later
// Author: Jasper Ford
// See NOTICE file for attribution and acknowledgements.

namespace App\Services;

use App\Core\TenantContext;
use App\Support\Safeguarding\SupportTiers;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Log;

/**
 * SupporterMessageViewService — a supporter's READ-ONLY window onto a
 * supported member's conversations (messages tier ≥ assist, consent-gated;
 * see SubAccountService::applyConsentedMessageAccess for how the tier rises).
 *
 * Design rules, each of which is a boundary test:
 *
 * - Fetches run AS THE MEMBER via MessageService, so the supporter sees
 *   exactly what the member sees: per-user delete filters, archiving, the
 *   lot. A message the member deleted for themselves is invisible here even
 *   though the other participant still has it.
 * - 🔴 NEVER marks anything read. markAsRead lives in MessagesController's
 *   own show() path, not in MessageService::getMessages — this service
 *   simply never calls it. The member's unread counts are their private
 *   state; a supporter's visit must leave no trace on them.
 * - `unread_count` is stripped from the list payload for the same reason.
 * - Federated conversations are excluded outright: the remote counterparty
 *   can never be shown the counterparty notice, so they can never be read
 *   under this capability.
 * - The safeguarding contact policy is re-asserted on EVERY read — a
 *   restriction landing after consent wins immediately.
 * - Every request writes one immutable audit row (mandatory purpose) BEFORE
 *   data is returned. No audit row, no data.
 */
class SupporterMessageViewService
{
    private array $errors = [];

    public function __construct(
        private readonly SubAccountService $subAccounts,
    ) {}

    public function getErrors(): array
    {
        return $this->errors;
    }

    /**
     * The supported member's conversation list, as they see it (minus their
     * private unread counts). Null with errors on refusal.
     *
     * @return array<int, array<string, mixed>>|null
     */
    public function listConversations(int $supporterUserId, int $supportedUserId, string $purpose, array $filters = []): ?array
    {
        if (! $this->authorize($supporterUserId, $supportedUserId, $purpose)) {
            return null;
        }

        $this->audit($supporterUserId, $supportedUserId, null, 'list', $purpose);

        // getConversations returns a paging envelope: ['items' => rows, 'cursor', 'has_more'].
        $payload = MessageService::getConversations($supportedUserId, $filters);

        // Rows can contain non-array entries (e.g. a partner row that failed
        // hydration maps to null) — drop them rather than crash.
        return array_values(array_map(static function (array $conversation): array {
            unset($conversation['unread_count']);
            return $conversation;
        }, array_filter($payload['items'] ?? [], 'is_array')));
    }

    /**
     * One thread, as the member sees it. Read-only: nothing is marked read.
     *
     * @return array<string, mixed>|null
     */
    public function viewThread(int $supporterUserId, int $supportedUserId, int $partnerUserId, string $purpose, array $filters = []): ?array
    {
        if (! $this->authorize($supporterUserId, $supportedUserId, $purpose)) {
            return null;
        }

        $this->audit($supporterUserId, $supportedUserId, $partnerUserId, 'read', $purpose);

        $payload = MessageService::getMessages($partnerUserId, $supportedUserId, $filters);
        if ($payload === null) {
            $this->errors[] = ['code' => 'NOT_FOUND', 'message' => __('api.user_not_found')];
            return null;
        }

        // Belt over MessageService's own filter: federated rows must never
        // surface here (their counterparty can never receive the notice).
        $payload['items'] = array_values(array_filter(
            $payload['items'] ?? [],
            static fn ($m) => ! (bool) (is_array($m) ? ($m['is_federated'] ?? false) : false),
        ));

        return $payload;
    }

    /**
     * The transparency counterpart: when the SUPPORTED MEMBER asks, tell them
     * when their supporter last looked and how often. Their own data — no
     * gate beyond being a party to the relationship.
     *
     * @return array{last_viewed_at: string|null, view_count_30d: int}
     */
    public function viewStatsForMember(int $supportedUserId, int $relationshipId): array
    {
        $row = DB::table('supporter_message_view_audits')
            ->where('tenant_id', TenantContext::getId())
            ->where('relationship_id', $relationshipId)
            ->where('supported_user_id', $supportedUserId)
            ->selectRaw('MAX(created_at) as last_viewed_at, SUM(created_at >= ?) as recent', [now()->subDays(30)])
            ->first();

        return [
            // Raw DATETIME has no timezone marker; browsers parse it as local
            // time and misreport the age. Emit ISO-8601 UTC instead.
            'last_viewed_at' => $row?->last_viewed_at
                ? \Illuminate\Support\Carbon::parse($row->last_viewed_at, 'UTC')->toIso8601String()
                : null,
            'view_count_30d' => (int) ($row?->recent ?? 0),
        ];
    }

    /**
     * Gate: purpose present, active relationship, messages tier ≥ assist
     * (tier vocabulary ONLY — hasPermission('can_view_messages') stays
     * hard-false forever), and the safeguarding contact policy re-asserted
     * at read time.
     */
    private function authorize(int $supporterUserId, int $supportedUserId, string $purpose): bool
    {
        $this->errors = [];

        if (trim($purpose) === '' || mb_strlen($purpose) > 500) {
            $this->errors[] = ['code' => 'VALIDATION_ERROR', 'message' => __('api.supporter_message_view_purpose_required')];
            return false;
        }

        $tiers = $this->subAccounts->resolvedTiers($supporterUserId, $supportedUserId);
        if ($tiers === null || ! SupportTiers::atLeast($tiers, 'messages', SupportTiers::ASSIST)) {
            $this->errors[] = ['code' => 'FORBIDDEN', 'message' => __('api_controllers_2.sub_account.no_permission')];
            return false;
        }

        try {
            app(SafeguardingInteractionPolicy::class)->assertLocalContactAllowed(
                $supporterUserId,
                $supportedUserId,
                TenantContext::getId(),
                'supporter_message_view',
            );
        } catch (\Throwable $e) {
            $this->errors[] = ['code' => 'FORBIDDEN', 'message' => $e->getMessage()];
            return false;
        }

        return true;
    }

    /**
     * One immutable row per request, written BEFORE data returns. A failure
     * here refuses the read: an unauditable view must not happen.
     */
    private function audit(int $supporterUserId, int $supportedUserId, ?int $partnerUserId, string $action, string $purpose): void
    {
        $relationshipId = (int) DB::table('account_relationships')
            ->where('tenant_id', TenantContext::getId())
            ->where('parent_user_id', $supporterUserId)
            ->where('child_user_id', $supportedUserId)
            ->where('status', 'active')
            ->value('id');

        DB::table('supporter_message_view_audits')->insert([
            'tenant_id' => TenantContext::getId(),
            'relationship_id' => $relationshipId,
            'supporter_user_id' => $supporterUserId,
            'supported_user_id' => $supportedUserId,
            'partner_user_id' => $partnerUserId,
            'action' => $action,
            'purpose' => mb_substr(trim($purpose), 0, 500),
            'correlation_hash' => hash('sha256', implode('|', [
                $supporterUserId, $supportedUserId, (string) $partnerUserId, trim($purpose), now()->toDateString(),
            ])),
            'created_at' => now(),
        ]);
    }
}

File: aspnet-backend/src/Nexus.Api/Services/SupporterMessageViewService.cs
// Copyright (c) 2024-2026 Jasper Ford
// SPDX-License-Identifier: AGPL-3.0-or-later
// Author: Jasper Ford
// See NOTICE file for attribution and acknowledgements.

using System.Security.Cryptography;
using System.Text;
using Microsoft.EntityFrameworkCore;
using Nexus.Api.Data;
using Nexus.Api.Entities;
using Nexus.Api.Support.Safeguarding;

namespace Nexus.Api.Services;

/// <summary>
/// Supervised message viewing — Laravel parity for
/// SupporterMessageViewService. A supporter holding the consented messages
/// tier (capped at assist) may LIST and READ the supported member's message
/// threads, and nothing else:
/// - a non-blank purpose (max 500 chars) is required on every view;
/// - the immutable audit row is written BEFORE any data is fetched;
/// - reads run AS the member: their deleted messages stay invisible;
/// - unread counts are stripped and nothing is ever marked read;
/// - there is deliberately NO write route under this surface.
/// </summary>
public class SupporterMessageViewService
{
    private readonly NexusDbContext _db;
    private readonly SafeguardingInteractionPolicy _safeguarding;
    private readonly ILogger<SupporterMessageViewService> _logger;
    private readonly List<object> _errors = [];

    public SupporterMessageViewService(
        NexusDbContext db,
        SafeguardingInteractionPolicy safeguarding,
        ILogger<SupporterMessageViewService> logger)
    {
        _db = db;
        _safeguarding = safeguarding;
        _logger = logger;
    }

    public IReadOnlyList<object> Errors => _errors;

    public async Task<object?> ListConversationsAsync(
        int supporterUserId, int supportedUserId, string? purpose, int limit,
        CancellationToken ct)
    {
        var relationship = await AuthorizeAsync(supporterUserId, supportedUserId, purpose, ct);
        if (relationship is null) return null;

        // Audit BEFORE fetching: there can never be a view without its record.
        await WriteAuditAsync(relationship, supporterUserId, supportedUserId,
            SupporterMessageViewAudit.ActionList, null, purpose!, ct);

        limit = Math.Max(1, Math.Min(limit, 50));
        var conversations = await _db.Conversations
            .AsNoTracking()
            .Where(c => c.Participant1Id == supportedUserId || c.Participant2Id == supportedUserId)
            .Select(c => new
            {
                c.Id,
                PartnerId = c.Participant1Id == supportedUserId ? c.Participant2Id : c.Participant1Id,
                c.CreatedAt,
                LastMessage = c.Messages
                    .Where(m => !m.IsDeleted
                        && (m.SenderId == supportedUserId ? !m.IsDeletedSender : !m.IsDeletedReceiver))
                    .OrderByDescending(m => m.Id)
                    .Select(m => new { m.Id, m.Content, m.CreatedAt })
                    .FirstOrDefault()
            })
            .ToListAsync(ct);

        var ordered = conversations
            .OrderByDescending(c => c.LastMessage?.Id ?? 0)
            .Take(limit + 1)
            .ToList();
        var hasMore = ordered.Count > limit;
        if (hasMore) ordered = ordered.Take(limit).ToList();

        var partnerIds = ordered.Select(c => c.PartnerId).Distinct().ToArray();
        var partners = await _db.Users.AsNoTracking()
            .Where(u => partnerIds.Contains(u.Id))
            .ToDictionaryAsync(u => u.Id, ct);

        var rows = ordered.Select(c => (object)new
        {
            partner_id = c.PartnerId,
            other_user = partners.TryGetValue(c.PartnerId, out var partner)
                ? new
                {
                    id = partner.Id,
                    name = $"{partner.FirstName} {partner.LastName}".Trim(),
                    first_name = partner.FirstName,
                    last_name = partner.LastName,
                    avatar_url = partner.AvatarUrl
                }
                : null,
            last_message = c.LastMessage is null
                ? null
                : new
                {
                    body = c.LastMessage.Content,
                    created_at = c.LastMessage.CreatedAt.ToString("yyyy-MM-dd'T'HH:mm:ssK")
                },
            created_at = c.CreatedAt.ToString("yyyy-MM-dd'T'HH:mm:ssK")
            // Deliberately NO unread_count: the supporter must never learn
            // what the member has or has not read.
        }).ToList();

        return new
        {
            conversations = rows,
            cursor = (string?)null,
            has_more = hasMore
        };
    }

    public async Task<object?> ShowThreadAsync(
        int supporterUserId, int supportedUserId, int partnerUserId, string? purpose, int limit,
        CancellationToken ct)
    {
        var relationship = await AuthorizeAsync(supporterUserId, supportedUserId, purpose, ct);
        if (relationship is null) return null;

        await WriteAuditAsync(relationship, supporterUserId, supportedUserId,
            SupporterMessageViewAudit.ActionRead, partnerUserId, purpose!, ct);

        var conversation = await _db.Conversations
            .AsNoTracking()
            .FirstOrDefaultAsync(c =>
                (c.Participant1Id == supportedUserId && c.Participant2Id == partnerUserId)
                || (c.Participant1Id == partnerUserId && c.Participant2Id == supportedUserId), ct);
        if (conversation is null)
        {
            _errors.Add(new { code = "NOT_FOUND", message = "Conversation not found" });
            return null;
        }

        limit = Math.Max(1, Math.Min(limit, 100));
        // Read AS the member: their per-message deletions stay deleted, and
        // nothing is ever marked read.
        var messages = await _db.Messages
            .AsNoTracking()
            .Where(m => m.ConversationId == conversation.Id && !m.IsDeleted)
            .Where(m => m.SenderId == supportedUserId ? !m.IsDeletedSender : !m.IsDeletedReceiver)
            .OrderByDescending(m => m.Id)
            .Take(limit)
            .ToListAsync(ct);
        messages.Reverse();

        var senderIds = messages.Select(m => m.SenderId).Distinct().ToArray();
        var senders = await _db.Users.AsNoTracking()
            .Where(u => senderIds.Contains(u.Id))
            .ToDictionaryAsync(u => u.Id, ct);

        return new
        {
            items = messages.Select(m => (object)new
            {
                id = m.Id,
                sender_id = m.SenderId,
                receiver_id = m.SenderId == supportedUserId ? partnerUserId : supportedUserId,
                body = m.Content,
                is_voice = false,
                created_at = m.CreatedAt.ToString("yyyy-MM-dd'T'HH:mm:ssK"),
                sender = senders.TryGetValue(m.SenderId, out var sender)
                    ? new
                    {
                        id = sender.Id,
                        first_name = sender.FirstName,
                        last_name = sender.LastName,
                        avatar_url = sender.AvatarUrl
                    }
                    : null
            }).ToList()
        };
    }

    private async Task<AccountRelationship?> AuthorizeAsync(
        int supporterUserId, int supportedUserId, string? purpose, CancellationToken ct)
    {
        _errors.Clear();
        if (string.IsNullOrWhiteSpace(purpose) || purpose.Trim().Length > 500)
        {
            _errors.Add(new
            {
                code = "VALIDATION_ERROR",
                message = "Say why you are viewing these messages"
            });
            return null;
        }

        var relationship = await _db.AccountRelationships
            .AsNoTracking()
            .FirstOrDefaultAsync(r => r.ParentUserId == supporterUserId
                && r.ChildUserId == supportedUserId
                && r.Status == AccountRelationship.StatusActive, ct);
        if (relationship is null
            || !SupportTiers.AtLeast(
                AccountRelationshipService.ResolvedTiers(relationship),
                "messages", SupportTiers.Assist))
        {
            _errors.Add(new
            {
                code = "FORBIDDEN",
                message = "You do not have permission to do this for that account"
            });
            return null;
        }

        try
        {
            // A safeguarding restriction beats an active grant.
            await _safeguarding.AssertLocalContactAllowedAsync(
                supporterUserId, supportedUserId, relationship.TenantId,
                "supporter_message_view", ct);
        }
        catch (Exception ex)
        {
            _errors.Add(new { code = "FORBIDDEN", message = ex.Message });
            return null;
        }

        return relationship;
    }

    /// <summary>The audit write is load-bearing: if it fails, the view fails.</summary>
    private async Task WriteAuditAsync(
        AccountRelationship relationship, int supporterUserId, int supportedUserId,
        string action, int? partnerUserId, string purpose, CancellationToken ct)
    {
        var trimmedPurpose = purpose.Trim();
        _db.SupporterMessageViewAudits.Add(new SupporterMessageViewAudit
        {
            TenantId = relationship.TenantId,
            RelationshipId = relationship.Id,
            SupporterUserId = supporterUserId,
            SupportedUserId = supportedUserId,
            PartnerUserId = partnerUserId,
            Action = action,
            Purpose = trimmedPurpose,
            CorrelationHash = Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(
                $"{relationship.TenantId}|{supporterUserId}|{supportedUserId}|{partnerUserId}|{trimmedPurpose}|{DateTime.UtcNow:yyyy-MM-dd}")))
                .ToLowerInvariant(),
            CreatedAt = DateTime.UtcNow
        });
        await _db.SaveChangesAsync(ct);
    }

}
